chia/full_node/mempool_manager.py

```diff
@@ -142,35 +143,82 @@ async def create_bundle_from_mempool(
-                    and item.fee + fee_sum <= self.constants.MAX_COIN_AMOUNT
-                ):
-                    spend_bundles.append(item.spend_bundle)
-                    cost_sum += item.cost
-                    fee_sum += item.fee
-                    removals.extend(item.removals)
-                    additions.extend(item.additions)
-                else:
-                    broke_from_inner_loop = True
-                    break
-        if len(spend_bundles) > 0:
+                consider_this_item = True
+                cost_saving = 0
+                spends_to_dedup: List[CoinSpend] = []
+                spends_to_run_for_cost: List[Tuple[bytes32, CoinSpend]] = []
+                # See if this item has coin spends that are eligible for deduplication
+                eligible_coin_ids = [s.coin_id for s in item.npc_result.conds.spends if s.eligible_for_dedup]
+                for coin_id in eligible_coin_ids:
+                    coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
+                    solution_hash_in_bundle = coin_spend_in_bundle.solution.get_tree_hash()
+                    # See if we processed an item with this coin before
+                    if coin_id in dedup_coin_spends:
+                        # See if the solution was identical
+                        processed_solution_hash, duplicate_cost = dedup_coin_spends[coin_id]
+                        if processed_solution_hash == solution_hash_in_bundle:
+                            # It was, so let's add this to the list of spends to deduplicate
+                            spends_to_dedup.append(coin_spend_in_bundle)
+                            cost_saving += duplicate_cost
+                        else:
+                            # It wasn't, so let's skip this whole transaction
+                            consider_this_item = False
+                            cost_saving = 0
+                            break
+                    else:
+                        # We didn't process an item with this coin before
+                        # If we end up including this transaction, add this item to
+                        # the ones we need to run in isolation and store the cost of
+                        spends_to_run_for_cost.append((coin_id, coin_spend_in_bundle))
+                if consider_this_item:
+                    # Run the spends that we're processing for the first time, for cost
+                    new_dedup_coin_spends: Dict[bytes32, Tuple[bytes32, uint64]] = {}
+                    for coin_id, coin_spend in spends_to_run_for_cost:
+                        spend_cost = uint64(coin_spend.puzzle_reveal.run_with_cost(item.cost, coin_spend.solution)[0])
+                        new_dedup_coin_spends[coin_id] = (coin_spend.solution.get_tree_hash(), spend_cost)
+                        cost_saving += spend_cost
```

Comment:
this is the first time we include this spend, isn't it? I don't see how this could give us any cost savings. This should be a unit test as well; having only one spend that is eligible for dedup (but doesn't have any duplicates). It should not affect the cost of the resulting bundle.